### ISSUE TITLE
Is there any way to match the performance of the paper?

### ISSUE THREAD

**Issue by wikilll** [Qznan/SpanKL#1]
When experimenting with SpanKL model using the training script in README.md, I found the test performance did not match the result of the paper. In Step 1, my Macro-F1 result in Split-All setup is 64.368%, and in Split-Filter is 66.489%. But in the paper, the results are 67.81% and 70.59%. Is there any change should I do to match the result?

**Comment by Qznan** [comment]
Thank you for the feedback！

Kindly note that _Tab.1_ and _Tab.2_ in the paper are the results **averaged over all the Task Permutations**. (Please refer to section _Metrics_ and _Appendix_). So, running the training script by default get results of one of them (i.e., --perm0). 

Actually, I once again tested the code to confirm results (in a new server) before reply, and finally got the **expected** result by calling `print_cl_metric()` after the training had finished, as follows:

```
# please run this before update the print_cl_metric.py 2023-03-03 :) !

import datautils as utils
from print_cl_metric import print_cl_metric

model_ckpt = 'fewnerd-0-2023-03-03_13-27-48-1406-spankl_split_perm0'
metrics_json = utils.load_json(f'model_ckpt/{model_ckpt}/overview_metric.json')
test_metric = metrics_json['test_metric']
filter_test_metric = metrics_json['filter_test_metric'] if 'filter_test_metric' in metrics_json else None
perm = model_ckpt.split('_')[-1]

print_cl_metric(metrics_json, model_ckpt, test_metric, perm, filter_test_metric)

# will output as belows:
======fewnerd======
task_best_dev_epo: [3, 4, 4, 3, 4, 4, 3, 3]
fewnerd-0-2023-03-03_13-27-48-1406-spankl_split_perm0

***Test All***
[[75.5   0.    0.    0.    0.    0.    0.    0.   75.5  75.5 ]
 [76.55 67.04  0.    0.    0.    0.    0.    0.   72.36 71.8 ]
 [76.57 67.17 61.8   0.    0.    0.    0.    0.   69.53 68.51]
 [76.64 67.06 62.51 63.09  0.    0.    0.    0.   68.91 67.33]
 [76.7  66.98 62.85 63.82 56.    0.    0.    0.   68.17 65.27]
 [76.69 66.69 62.68 63.87 56.96 61.96  0.    0.   67.79 64.81]
 [76.75 66.31 62.88 64.11 56.53 63.06 60.32  0.   67.51 64.28]
 [76.5  66.26 62.59 63.98 56.77 63.32 60.47 55.03 66.88 63.11]
 [-0.26 -0.91 -0.29 -0.14 -0.19  0.    0.    0.   -0.22  0.  ]]

***Test Filter***
[[77.7   0.    0.    0.    0.    0.    0.    0.   77.7  77.7 ]
 [77.86 68.17  0.    0.    0.    0.    0.    0.   73.59 73.02]
 [77.27 67.78 62.81  0.    0.    0.    0.    0.   70.3  69.29]
 [77.26 67.42 63.42 64.67  0.    0.    0.    0.   69.66 68.19]
 [77.24 67.31 63.57 65.21 56.69  0.    0.    0.   68.82 66.01]
 [77.   67.01 63.23 65.29 57.49 62.21  0.    0.   68.3  65.37]
 [77.04 66.56 63.29 65.5  57.   63.29 60.52  0.   67.95 64.74]
 [76.7  66.49 62.89 65.41 57.25 63.52 60.59 55.28 67.25 63.51]
 [-1.17 -1.68 -0.69 -0.09 -0.24  0.    0.    0.   -0.48  0.  ]]
```
The last column of the above tables is the metrics used in the paper, i.e., the expected results of --perm0. (Note that _Fig.3_ in the paper is also the results of --perm0)

In addition, note that Few-NERD asks model to learn multiple entity types per task, and we use the **microF1** over the entity types inner a task as this task's metric. For every step in CL, we use **macroF1** over all the learned tasks as the final metric (refer to section _Metrics_), for example: 
Step1 is Micro(Task1)
Step2 is Macro( Micro(Task1), Micro(Task2))
Step3 is Macro( Micro(Task1), Micro(Task2), Micro(Task3))
...

All the models conducting Few-NERD experiment comply with this principle.

So, some metrics printed in the middle of the training log (typically Few-NERD) are **not** the results listed in the paper, they are the micro/macroF1 considering all fine-grained entities equally, for reference only.

And we will newly commit to also consider timely printing the results by `print_cl_metric()` during training to reduce possible misunderstanding.

Hope the reply can help you!

**Comment by wikilll** [comment]
I re-run the training code and use print_cl_metric() to get the results. My confusion is settled.

Thank you very much for your patience and detailed explanation!
